File: pset6/credit/credit.py
import cs50

card = cs50.get_string("Number: ")

sum = 0
card_odd = card[len(card)-2::-2]
card_even = card[::-2]

for odd in card_odd:
    tmp = int(odd) * 2
    for i in str(tmp):
        sum += int(i)

for even in card_even:
    sum += int(even)

if sum % 10 == 0:
    if len(card) == 15 and (card[0:2] in ['34', '37']):
        print("AMEX")
    elif len(card) == 16 and (card[0:2] in ['51', '52', '53', '54', '55']):
        print("MASTERCARD")
    elif len(card) in [13, 16] and card[0] == '4':
        print("VISA")
    else:
        print("INVALID")

else:
    print("INVALID")

File: pset1/credit/credit.c
#include<stdio.h>
#include<cs50.h>

int main(void)
{
    bool odd = true;
    int sumOdd = 0;
    int sumEven = 0;
    bool isAmex = false, isVisa = false, isMC = false;
    int length = 0;
    long creditCard = get_long("Number: ");
    while (creditCard != 0)
    {
        // check each digit and add sum
        if (odd)
        {
            sumOdd += creditCard % 10;
            odd = !odd;
        }
        else
        {
            int tmp = creditCard % 10;
            tmp *= 2;
            sumEven = sumEven + tmp % 10 + tmp / 10;
            odd = !odd;
        }

        // check the first two or the first digits and determine card type
        if (creditCard >= 51 && creditCard <= 55)
        {
            isMC = true;
        }
        else if (creditCard == 34 || creditCard == 37)
        {
            isAmex = true;
        }
        else if (creditCard == 4)
        {
            isVisa = true;
        }

        // go to next digit
        creditCard = creditCard / 10;
        length ++;
    }
    int sumAll = sumOdd + sumEven;
    if (sumAll % 10 != 0)
    {
        // invalid card type
        printf("INVALID\n");
    }
    else
    {
        // valid card type
        if (isMC && length == 16)
        {
            printf("MASTERCARD\n");
        }
        else if (isVisa && (length == 13 || length == 16))
        {
            printf("VISA\n");
        }
        else if (isAmex && length == 15)
        {
            printf("AMEX\n");
        }
        else
        {
            printf("INVALID\n");
        }
    }
}